### PR TITLE
cleanup(tests,virtctl): Refactor and cleanup imageupload functests

### PR DESCRIPTION
**Pull request by 0xFelix**
### What this PR does

Refactor and cleanup the imageupload functests by:

- Using tables where possible and adding missing test cases
- Removing the storage-req decorator as it looks to be not required
- Avoiding the need for the tests to run in Serial
- Avoiding lambdas where not required
- Simplifying helper functions
- Dropping test cases that are covered by unit tests:
  - [test_id:10674]Should fail with invalid volume-mode
  - Upload fails creating a DV when using a non-existent storageClass
  - Upload fails creating a PVC when using a non-existent storageClass

Change the existing unit test that verifies an upload fails on using a
non existent storageClass to be a Table, so it covers both DVs and PVCs.
This was previously a functional test.

To overcome situations where the upload server Pod might not be ready
yet, add a retry to the virtctl imageupload command, that retries up to 15 times.

Before this PR:

virtctl image-upload functests run in Serial and duplicate unit tests

After this PR:

virtctl image-upload functests run in Parallel and do not duplicate unit tests

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Special notes for your reviewer

Wait for #12880 to merge first

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
virtctl: Image uploads are retried up to 15 times
```

### CODE REVIEW

File: pkg/virtctl/imageupload/imageupload_test.go [hunk 1]
```diff
@@ -899,19 +899,30 @@ var _ = Describe("ImageUpload", func() {
 			Expect(cmd()).NotTo(Succeed())
 		})
 
-		It("Upload fails when using a nonexistent storageClass", func() {
+		DescribeTable("Upload fails when using a nonexistent storageClass", func(resource string) {
+			const errFmt = "storageclasses.storage.k8s.io \"%s\" not found"
```
Comment (0xFelix) [inline pkg/virtctl/imageupload/imageupload_test.go:903]: Make it a const block

**Review by 0xFelix — COMMENTED** [review]
Reword commit refactor -> cleanup

**Comment by 0xFelix** [comment]
@mhenriks @awels Can you help me find out what is causing the test flakes?

> `unexpected return value 502, error in upload-proxy: http: proxy error: dial tcp 10.106.212.29:443: connect: connection refused`

Looks like the upload proxy is unavailable from time to time?

**Comment by 0xFelix** [comment]
The issue looks to be CDI marking the upload Pod as ready too fast: https://github.com/kubevirt/containerized-data-importer/issues/3450

**Comment by 0xFelix** [comment]
@mhenriks @awels Can you please take a look at this?

**Comment by 0xFelix** [comment]
@mhenriks @awels Do you want me to quarantine the flaky tests?

**Comment by 0xFelix** [comment]
Ping @mhenriks Can you PTAL?

**Comment by mhenriks** [comment]
@0xFelix apoligies for the delay, looks good, rebase and I will take another look

**Comment by 0xFelix** [comment]
@mhenriks I rebased the PR, but I expect the flakes to still be an issue.

I saw your https://github.com/kubevirt/containerized-data-importer/issues/3450#issuecomment-2426515784. I need to have a look again this.

File: pkg/virtctl/imageupload/imageupload.go [hunk 2]
```diff
@@ -439,6 +440,8 @@ func ConstructUploadProxyPathAsync(uploadProxyURL, token string, insecure bool)
 }
 
 func (c *command) uploadData(token string, file *os.File) error {
+	const maxRetries = 15
```
Comment (mhenriks) [inline pkg/virtctl/imageupload/imageupload.go:443]: Should we do what we did with export and make this a proper arg with a default of 2?

Comment (0xFelix) [reply pkg/virtctl/imageupload/imageupload.go]: ACK

**Comment by jcanocan** [comment]
Nice job!
Nothing to add in my side.
/lgtm

**Comment by akalenyu** [comment]
So a while ago I experimented with running the imageupload tests in parallel, and it didn't go very well:
https://github.com/kubevirt/kubevirt/pull/11716#issuecomment-2373256509
I am pretty sure this is also why this PR is introducing flaking in the imageupload tests:
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/12800/pull-kubevirt-e2e-k8s-1.29-sig-storage/1857478378622291968

I will open a PR to put them back to Serial for now as the gain in runtime is not very drastic

**Pull request merged**